compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -584,51 +832,119 @@ void PropagateLinalgTransposePass::runOnOperation() {
-                          /*ReplacementType=*/linalg::BatchMatmulTransposeAOp,
-                          /*inputIdx=*/0>>(context,
-                                           SmallVector<int64_t>{0, 2, 1});
-    sinkingPatterns
-        .insert<NamedOpConversion</*OpType=*/linalg::BatchMatmulTransposeBOp,
-                                  /*ReplacementType=*/linalg::BatchMatmulOp,
-                                  /*inputIdx=*/1>>(
-            context, SmallVector<int64_t>{0, 2, 1});
-    sinkingPatterns
-        .insert<NamedOpConversion</*OpType=*/linalg::BatchMatmulTransposeAOp,
-                                  /*ReplacementType=*/linalg::BatchMatmulOp,
-                                  /*inputIdx=*/0>>(
-            context, SmallVector<int64_t>{0, 2, 1});
     sinkingPatterns.insert<SinkTransposeThroughExtractSlice>(context);
     sinkingPatterns.insert<SinkTransposeThroughExpandShape>(context);
-    sinkingPatterns.insert<FuseTransposeWithGenericConsumer>(context);
+    populateNamedOpSinkingPatterns(context, sinkingPatterns);
     sinkingPatterns.add<PropagateTransposeThroughUnaryElementwiseInput>(
         context, /*benefit=*/2);
+    if (failed(
+            applyPatternsAndFoldGreedily(funcOp, std::move(sinkingPatterns)))) {
+      return signalPassFailure();
+    }
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "\n--- After canonicalizing transpose in place ---\n";
+    funcOp->print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Propagate transposes upwards, and fuse with any producer generic ops. Also
+  // propagate reshapes upwards to open up more transpose fusion opportunities.
+  if (!testSinkingOnly) {
+    linalg::ControlFusionFn reshapePropagationFn = [](OpOperand *fusedOperand) {
+      Operation *producer = fusedOperand->get().getDefiningOp();
+      Operation *consumer = fusedOperand->getOwner();
+      if (!IREE::Flow::isNonNullAndOutsideDispatch({producer, consumer})) {
+        return false;
+      }
+
+      // Do not reshape producer linalg op if it has more than one user.
+      auto producerLinalgOp = dyn_cast<linalg::LinalgOp>(producer);
+      if (!producerLinalgOp || !producerLinalgOp->hasOneUse()) {
+        return false;
+      }
+      // Only reshape generic ops and contraction ops.
+      if (!linalg::isaContractionOpInterface(producerLinalgOp) &&
+          !isa<linalg::GenericOp>(producerLinalgOp)) {
+        return false;
+      }
+      // Only propagate reshapes up through generics in this case, hoping
+      // to increase the chance we can fuse with a nearby transpose.
+      if (!isa<tensor::ExpandShapeOp>(consumer) &&
+          !isa<tensor::CollapseShapeOp>(consumer)) {
+        return false;
+      }
+
+      return consumer->hasOneUse() &&
+             llvm::isa<linalg::TransposeOp>(*(consumer->user_begin()));
+    };
+    RewritePatternSet bubblingPatterns(context);
+    linalg::populateFoldReshapeOpsByExpansionPatterns(bubblingPatterns,
```

Comment:
If reshapes are blocking things then we should run these before doing transpose propagation. Also note that the reshape by expansion and reshape by collapse are converse of each other. If you put both together and use the same control function that seems problematic to me

Reply:
The control function here is to "only do it if the reshape (collapse or expand) is a consumer of the generic." So basically this will never loop infinitely as it only attempts to move reshapes up the SSA chain (and vice versa for pushing reshapes down the SSA chain).

We might need to restrict this in cases where the generic has multiple results.